src/foam/box/README.md

```diff
@@ -0,0 +1,211 @@
+# Boxes
+
+The `foam.box.Box` interface is a message-passing interface. Its various
+implementations facilitate message passing over various boundaries such as
+FOAM contexts, operating system processes, and network layers.
+
+## Boxes and messages
+
+The box interface contains one method:
+
+```
+interface Box {
+  void send(message);
+}
+```
+
+The message interface contains a payload and optional routing information:
+
+```
+interface Message {
+  Object object;     // payload
+  Object attributes; // routing information
+}
+```
+
+## Some patterns over boxes
+
+The Box library contains base classes for several patterns.
+
+### Proxy
+
+A `ProxyBox` is simply any box with a `delegate` of type `Box`. The base
+`ProxyBox` simply forwards messages to its `delegate`. Other implementations
+typically modify input messages, and then pass the modified message to
+their `delegate`.
+
+### Promised
+
+A `PromisedBox` has a `delegate` property that can be set to a
+`Promise<Box>`. The `PromisedBox` itself will accept messages immediately and
+deliver them when its `delegate` is resolved.
+
+## Boxes as named services
+
+Boxes are typically used to register and expose services that accept
+messages. Registering and routing to such services is the role of the
+`BoxRegistry` interface:
+
+```
+interface BoxRegistry {
+  Box  register(opt_name, opt_clientServerService, box);
+  void unregister(nameOrBox);
+}
+```
+
+The `register` method is used to register the input `box` under the name
+`opt_name`. If `opt_name` is omitted, the registry selects an available
```

Comment:
The default implementation is not random. It uses a counter. https://github.com/foam-framework/foam2/blob/master/src/lib/box.js#L235